src/runtime/singlethreaded.rs

```diff
@@ -0,0 +1,83 @@
+use executor::current_thread::{self, CurrentThread};
+
+use tokio_reactor::{self, Reactor};
+use tokio_timer::timer::{self, Timer};
+use tokio_executor;
+
+use futures::Future;
+
+use std::io;
+
+/// Single-threaded runtime provides a way to start reactor
+/// and executor on the same thread
+///
+/// # Examples
+///
+/// Creating a new `Runtime` with default configuration values.
+///
+/// ```
+/// use tokio::runtime::SingleThreaded;
+/// use tokio::prelude::*;
+///
+/// let rt = SingleThreaded::new().unwrap();
+///
+/// // Use the runtime...
+/// // rt.block_on(f); // where f is a future
+/// ```
+#[derive(Debug)]
+pub struct SingleThreaded {
+    reactor_handle: tokio_reactor::Handle,
+    timer_handle: timer::Handle,
+    executor: CurrentThread<Timer<Reactor>>,
+}
+
+impl SingleThreaded {
+    /// Returns a new single-threaded runtime initialized with default
+    /// configuration values.
+    pub fn new() -> io::Result<SingleThreaded> {
+        // We need a reactor to receive events about IO objects from kernel
+        let reactor = Reactor::new()?;
+        let reactor_handle = reactor.handle();
+
+        // Place a timer wheel on top of the reactor. If there are no timeouts to fire, it'll let the
+        // reactor pick up new some new external events.
+        let timer = Timer::new(reactor);
+        let timer_handle = timer.handle();
+
+        // And now put a single-threaded executor on top of the timer. When there are no futures ready
+        // to do something, it'll let the timer or the reactor to generate some new stimuli for the
+        // futures to continue in their life.
+        let executor = CurrentThread::new_with_park(timer);
+
+        let runtime = SingleThreaded { reactor_handle, timer_handle, executor };
+        Ok(runtime)
+    }
+
+    /// Blocks and runs the given future.
+    ///
+    /// This is similar to running a runtime, but uses only the current thread.
+    pub fn block_on<F: Future<Item = (), Error = ()>>(&mut self, f: F) -> () {
+        let reactor_handle = self.reactor_handle.clone();
```

Comment:
Are the clones necessary? They are only passed by reference to the fns below.

Reply:
```
error[E0500]: closure requires unique access to `self` but `self.reactor_handle` is already borrowed
  --> src/runtime/singlethreaded.rs:64:71
   |
64 |         tokio_reactor::with_default(&self.reactor_handle, &mut enter, |enter| {
   |                                      -------------------              ^^^^^^^ closure construction occurs here
   |                                      |
   |                                      borrow occurs here
65 |             timer::with_default(&self.timer_handle, enter, |enter| {
   |                                  ---- borrow occurs due to use of `self` in closure
...
79 |         });
   |          - borrow ends here

error[E0500]: closure requires unique access to `*self` but `self.timer_handle` is already borrowed
  --> src/runtime/singlethreaded.rs:65:60
   |
65 |             timer::with_default(&self.timer_handle, enter, |enter| {
   |                                  -----------------         ^^^^^^^ closure construction occurs here
   |                                  |
   |                                  borrow occurs here
...
72 |                     let mut executor = self.executor.enter(enter);
   |                                        ---- borrow occurs due to use of `*self` in closure
...
78 |             });
   |              - borrow ends here

error: aborting due to 2 previous errors

```

Reply:
Outside the closure you could borrow all the things you need at once if I'm reading the problem correctly, e.g.
```rust
let SingleThreaded { ref reactor_handle, ref timer_handler, .. } = &self;
```

Reply:
@sdroege thanks.
@carllerche done.